### PR TITLE
[ADD] web_model_read_only

### PR DESCRIPTION
**Pull request by ilyasProgrammer**
Module disables modification operations (write, create, delete) on selected model to selected users groups.

### CODE REVIEW

**Comment by yostashiro** [comment]
@ilyasProgrammer Not sure your business requirement but could https://github.com/OCA/server-tools/tree/14.0/base_model_restrict_update cover your use case?

**Comment by francesco-ooops** [comment]
Hi @yostashiro, thanks for the heads up! I'm a functional and I was testing your module, if I understand correctly: 

- you can set one or more models to be edited only by "permitted" users

- "permitted" users can edit all restricted models?

I ask as I tried to test on runbot but couldn't find either "Grant Update Permit" or "Readonly User" in user form.

**Comment by yostashiro** [comment]
> * you can set one or more models to be edited only by "permitted" users
> * "permitted" users can edit all restricted models?

@francesco-ooops Yes for both.

> I ask as I tried to test on runbot but couldn't find either "Grant Update Permit" or "Readonly User" in user form.

I realize that the button presentation in the user form has been broken and created a PR to fix it here: https://github.com/OCA/server-tools/pull/2626

**Comment by francesco-ooops** [comment]
Hi @yostashiro , 

your solution works as described but has two limitations IMHO: 

- I cannot allow User A to edit Model 1 but not Model 2, and User B to edit Model 2 but not Model 1

- If I have 10 users allowed to edit Model 1 and I need to restrict one of those users from editing it, I need to set 9 users as "permitted", which is a bit counterintuitive

On the other hand, I like that the flag "read only" sets virtually the whole DB in "guest mode", although the "read only" implemented only restricts user from saving records, but allows for example to browse related objects (eg: read-only user in SO cannot change customer, but can check all customer records)

I think the solution proposed in this module is different enough to allow both modules to be published in OCA :)

**Review by francesco-ooops — APPROVED** [review]
Functional review ok!

**Comment by francesco-ooops** [comment]
@yostashiro what do you think?

File: web_model_read_only/models/models.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    )
+
+
+class BaseModel(models.AbstractModel):
+    _inherit = "base"
+
+    @api.model
+    def check_access_rights(self, operation, raise_exception=True):
+        if operation != "read":
+            model = self.env["ir.model"].search([("model", "=", self._name)])
```
Comment (geomer198) [inline web_model_read_only/models/models.py]: You could use `_get` function get model record by model name.
For example:
```
self.env['ir.model']._get(self._name)
```

**Review by geomer198 — COMMENTED** [review]
Please add test coverage for functional.

**Comment by yostashiro** [comment]
@francesco-ooops I see that the design concept of this module is somewhat similar to that of https://github.com/OCA/server-ux/tree/14.0/base_menu_visibility_restriction - normally Odoo makes things more permissive for groups in the upper node, but the design of this module works in the opposite way.  Therefore, I suppose that you are naturally expected to define new groups to be assigned to users and models, specifically for the purpose of access restriction.

I think the function can be useful as long as the usage is properly documented.

I was thinking, however, it may still be good to conflate the proposed function into base_model_restrict_update.  Maybe we can think about it for future versions.

Shouldn't this module be put in https://github.com/OCA/server-tools instead?

**Review by geomer198 — CHANGES_REQUESTED** [review]
Please add tests for functional.

File: web_model_read_only/tests/test_models.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+from odoo.tests.common import SavepointCase
+
+
+class TestIrModel(SavepointCase):
+    @classmethod
+    def setUpClass(cls):
+        super(TestIrModel, cls).setUpClass()
+        cls.readonly_group = cls.env.ref("base.group_user")
+        cls.group_manager = cls.env.ref("base.group_erp_manager")
+        cls.ir_model = cls.env["ir.model"].create(
```
Comment (geomer198) [inline web_model_read_only/tests/test_models.py]: For what you create new model? Could you use existing model for example `res.partner`?

File: web_model_read_only/tests/test_models.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            {"name": "x_test_model", "model": "x_test.model"}
+        )
+        User = cls.env["res.users"]
```
Comment (geomer198) [inline web_model_read_only/tests/test_models.py]: Better change naming. For this situation and similar use: `res_users obj`. `obj` suffix used for the showing what variable contain empty model object.

File: web_model_read_only/tests/test_models.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        )
+        cls.user_readonly = User.create(
+            {
+                "name": "User readonly",
+                "login": "user_readonly",
+                "groups_id": cls.readonly_group,
+            }
+        )
+
+    def test_check_access_rights(self):
```
Comment (geomer198) [inline web_model_read_only/tests/test_models.py]: Please split this function on the separate cases.
And will make more tests for the coverage all cases.

**Review by geomer198 — CHANGES_REQUESTED** [review]
Please fix comments.

**Review by geomer198 — COMMENTED** [review]
For all assertion check need will add message.

**Comment by francesco-ooops** [comment]
@ilyasProgrammer can you close this one?

**Pull request closed without merging**